### PR TITLE
[OTLP/GRPC] Ensure OTLP receiver handles consume errors correctly 

### PR DESCRIPTION
**Pull request by VihasMakwana**
**Description:** Follow the receiver contract and return Unavailable for non-permanent and InvalidArgument for permanent errors for OTLP/gRPC receiver.

Leave the "Retry-After" field blank and let the client implement an exponential backoff strategy.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/4335

**Testing:** Added relevant test cases.

### CODE REVIEW

**Comment by VihasMakwana** [comment]
@bogdandrutu can you have a quick look at this, if it's fine.

File: receiver/otlpreceiver/internal/metrics/otlp.go
Comment (evan-bradley) [inline receiver/otlpreceiver/internal/metrics/otlp.go:47]: After looking at https://github.com/open-telemetry/opentelemetry-collector/pull/8038, we will need to reconsider our overall approach. `status.Status` doesn't have an `Unwrap` method, so `consumererror.IsPermanent` will always be false when called on errors returned from this function.

The way I see it, we can:
1. Translate gRPC status codes to HTTP status codes in the OTLP/HTTP receiver
2. Perform this logic in a gRPC-only wrapper function similar to what the OTLP/HTTP receiver has.

What do you think?

Comment (VihasMakwana) [reply receiver/otlpreceiver/internal/metrics/otlp.go:47]: That's exactly what I'm planning to do once I get this merged. And that's why I put https://github.com/open-telemetry/opentelemetry-collector/pull/8038 to draft mode.

Comment (evan-bradley) [reply receiver/otlpreceiver/internal/metrics/otlp.go:47]: >That's exactly what I'm planning to do once I get this merged.

I'm sorry, I'm not clear on what you mean here. Are you saying you plan to update https://github.com/open-telemetry/opentelemetry-collector/pull/8038 to do gRPC->HTTP status code translation?

Comment (VihasMakwana) [reply receiver/otlpreceiver/internal/metrics/otlp.go:47]: yeah, that's correct.

File: receiver/otlpreceiver/internal/logs/otlp_test.go
Comment (VihasMakwana) [reply receiver/otlpreceiver/internal/logs/otlp_test.go:57]: It would be `status.Error`, as we do `s.Err()` in `Export`. 
I will add a check for that.

File: receiver/otlpreceiver/otlp_test.go
Comment (VihasMakwana) [reply receiver/otlpreceiver/otlp_test.go:696]: Done

**Comment by VihasMakwana** [comment]
@bogdandrutu @evan-bradley ^^

File: receiver/otlpreceiver/internal/metrics/otlp.go
Comment (evan-bradley) [inline receiver/otlpreceiver/internal/metrics/otlp.go:58]:
```suggestion
	return pmetricotlp.NewExportResponse(), nil
```

File: receiver/otlpreceiver/internal/trace/otlp.go
Comment (evan-bradley) [inline receiver/otlpreceiver/internal/trace/otlp.go:59]:
```suggestion

	return ptraceotlp.NewExportResponse(), nil
```

**Review by evan-bradley — APPROVED** [review]

**Comment by VihasMakwana** [comment]
@bogdandrutu can you a quick look if it's fine by you?

**Comment by evan-bradley** [comment]
Bogdan will be out for a few more weeks. @open-telemetry/collector-approvers could one of you please take a look?

**Comment by VihasMakwana** [comment]
not stale, waiting for review

**Comment by VihasMakwana** [comment]
@evan-bradley do you happen to know when @bogdandrutu will be back?

**Comment by evan-bradley** [comment]
He is back, but is likely still catching up. @open-telemetry/collector-approvers would one of you be available to take a look?

@VihasMakwana it looks like a few tests are failing, could you rebase on main?

**Comment by VihasMakwana** [comment]
@bogdandrutu @codeboten the CI's green, can one of you take a look, please?

**Comment by VihasMakwana** [comment]
@dmitryax would you be available to take a look?

**Comment by VihasMakwana** [comment]
waiting for review on @open-telemetry/collector-approvers. not stale

File: receiver/otlpreceiver/internal/logs/otlp.go
Comment (atoulme) [inline receiver/otlpreceiver/internal/logs/otlp.go:47]: is there a reference to where this is defined? The comment seems to describe what the code attempts to do, but it'd be great to have a link or reference to how this should be, for people like me who know nothing

File: receiver/otlpreceiver/internal/metrics/otlp.go
Comment (atoulme) [inline receiver/otlpreceiver/internal/metrics/otlp.go:47]: same comment as above

File: receiver/otlpreceiver/internal/logs/otlp.go
Comment (VihasMakwana) [reply receiver/otlpreceiver/internal/logs/otlp.go:47]: @atoulme can you check now, I've added a reference.

Comment (atoulme) [reply receiver/otlpreceiver/internal/logs/otlp.go:47]: Perfect thanks!

**Comment by VihasMakwana** [comment]
@bogdandrutu can we get this merged? https://github.com/open-telemetry/opentelemetry-collector/issues/8676 seems to waiting on it.

**Review by atoulme — APPROVED** [review]

**Review by yurishkuro — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]
LGTM. 

@open-telemetry/collector-approvers I will merge this in 5 working days unless someone blocks it

File: receiver/otlpreceiver/internal/trace/otlp.go
Comment (TylerHelmuth) [inline receiver/otlpreceiver/internal/trace/otlp.go:62]: Since this error handling is the same for all signals I do wonder if it could be extracted and re-used, but that is an optimization we could handle later.

**Review by TylerHelmuth — APPROVED** [review]

File: receiver/otlpreceiver/internal/trace/otlp.go
Comment (mx-psi) [reply receiver/otlpreceiver/internal/trace/otlp.go:62]: Filed #9300 for this

**Pull request merged**
